qiskit/synthesis/evolution/product_formula.py

```diff
@@ -239,3 +250,74 @@ def real_or_fail(value, tol=100):
+
+
+def reorder_paulis(
+    paulis: SparsePauliLabel, strategy: rx.ColoringStrategy = rx.ColoringStrategy.Saturation
+) -> SparsePauliOp | list[SparsePauliOp]:
+    r"""
+    Creates an equivalent operator by preserve_ordering terms in order to yield a
+    shallower circuit after evolution synthesis. The original operator remains
+    unchanged.
+
+    This method works in three steps. First, a graph is constructed, where the
+    nodes are the terms of the operator and where two nodes are connected if
+    their term acts on the same qubit (for example, the terms :math:`IXX` and
+    :math:`IYI` would be connected, but not :math:`IXX` and :math:`YII`). Then,
+    the graph is colored.  Two terms with the same color thus do not act on the
+    same qubit, and in particular, their evolution subcircuits can be run in
+    parallel in the greater evolution circuit of ``operator``. Finally, a new
+    :class:`~qiskit.quantum_info.SparsePauliOp` is created where terms of the
+    same color are grouped together.
+
+    In trivial cases, i.e. when either
+    - the input is a :class:`~qiskit.quantum_info.SparsePauliOp` with
+      less than two Pauli terms, or
+    - the input is a list containing a single
+      :class:`~qiskit.quantum_info.SparsePauliOp` which has less than two
+      Pauli terms,
+    this method does nothing.
+
+    If ``operators`` is a list of :class:`~qiskit.quantum_info.SparsePauliOp`,
+    then preserve_ordering is applied to every operator independently, and the list of
+    preserve_ordered operators is returned.
+
+    This method is deterministic and invariant under permutation of the Pauli
+    term in ``operators``.
+
+    Args:
+        paulis: The operator whose terms to preserve_order.
+        strategy: The coloring heuristic to use, see ``ColoringStrategy`` [#].
+            Default is ``ColoringStrategy.Saturation``.
+
+    .. [#] https://www.rustworkx.org/apiref/rustworkx.ColoringStrategy.html#coloringstrategy
+
+    """
+
+    def _term_sort_key(term: SparsePauliLabel) -> typing.Any:
+        # sort by index, then by pauli
+        return (term[1], term[0])
+
+    # Do nothing in trivial cases
+    if len(paulis) <= 1:
+        return paulis
+
+    terms = sorted(paulis, key=_term_sort_key)
+    graph = rx.PyGraph()
+    graph.add_nodes_from(terms)
+    indexed_nodes = list(enumerate(graph.nodes()))
+    for (idx1, (_, ind1, _)), (idx2, (_, ind2, _)) in combinations(indexed_nodes, 2):
+        # Add an edge between two terms if they touch the same qubit
+        if len(set(ind1).intersection(ind2)) > 0:
+            graph.add_edge(idx1, idx2, None)
+
+    # rx.graph_greedy_color is supposed to be deterministic
+    coloring = rx.graph_greedy_color(graph, strategy=strategy)
+    terms_by_color = defaultdict(list)
+
+    for term_idx, color in sorted(coloring.items()):
+        term = graph.nodes()[term_idx]
+        terms_by_color[color].append(term)
+
+    terms = sum(terms_by_color.values(), [])
```

Comment:
This is probably obvious, but how is `sum` used here exactly? And why at the end `terms` is of the form  `SparsePauliOp | list[SparsePauliOp]`?

Reply:
It misuses `sum` to concatenate lists -- it's better to use `itertools.chain` though, I'll update

Reply:
Fixed in 56e619c